docs/WitnessMatchingManifesto.md

```diff
@@ -0,0 +1,273 @@
+- A _protocol witness_ (or just _witness_) is a value or a type that satisfies a protocol requirement.
+
+For example:
+
+```swift
+protocol P {
+  func foo() // This is a protocol requirement
+}
+
+struct S: P {
+  func foo() { // This is a protocol witness
+    ...
+  }
+}
+```
+
+## Introduction
+
+In Swift (at the time of this writing), protocol witnesses have to pretty much match exactly with a protocol requirement. For example:
+
+```swift
+protocol P {
+  func foo()
+}
+
+struct S1: P {
+  func foo() { ... } // This matches
+}
+
+struct S2: P {
+  func foo(arg: Int = 0) { ... } // This is not a match
+}
+```
+
+```swift
+protocol Q {
+  func getView() -> UIView
+}
+
+struct S1: Q {
+  func getView() -> UIView { ... } // This matches
+}
+
+struct S2: Q {
+  func getView() -> MyCustomUIViewSubclass {  // This is not a match 
+    ...
+  }
+}
+```
+
+However, there are some exceptions to these rules, for example:
+
+1. A non-throwing function can witness a throwing function requirement:
+
+```swift
+protocol P {
+  func foo() throws
+}
+
+struct S: P {
+  // Even though 'foo' here is not marked 'throws', it is
+  // still able to satisfy the protocol requirement. The
+  // caller has to handle the possibility that it might 
+  // throw, but the witness is ofcourse free to throw or 
+  // not to, depending on whether it makes sense in the
+  // implementation.
+  func foo() { ... }
+}
+```
+
+2. A non-failable initializer can witness a failable initializer requirement:
+
+```swift
+protocol P {
+  init?()
+}
+
+struct S: P {
+  // Similarly, the initializer here is free to return
+  // a newly created value or just return 'nil' instead.
+  init() {}
+}
+```
+
+3. A non-escaping closure can witness one that is marked `@escaping`:
+
+```swift
+protocol P {
+  func foo(completion: @escaping() -> ()) 
+}
+
+struct S: P {
+  // Similar to above, you can use a non-escaping closure here
+  // if you want.
+  func foo(completion: () -> ()) {
+    ...
+  }
+}
+```
+
+Some of these rules seem reasonable, but one might argue that the existing model is just too restrictive in general (for example, subtyping is not allowed here, but is allowed in classes). This leads us to ask a very basic question - what kinds of differences are reasonable when matching a protocol witness to a protocol requirement?
+
+This manifesto lists some possible changes we can make to the current model and set out some basic design principles. Some of these changes have been discussed many times in the past on the forums and while there is no guarantee that all of these changes will be implemented (either now or in the future), this document will serve as a  starting point for discussions around witness matching (such as discussions around  potential designs of how a certain feature mentioned here should work, whether its the right thing to do, etc) and something the community can use to create and publish evolution proposals.
```

Comment:
This paragraph should set an agenda for the rest of the document; it's essentially a table of contents in prose.  So how about something more like:

> This manifesto starts by listing some possible changes we could make that would allow more mismatches.  It then sets out some design principles and describes some of the constraints on how the language can reasonably behave.  Finally, it proposes a basic model for how witness matching should work in the language.  The ultimate goal of this document is to provide a foundation for discussing witness matching in Swift and to spur a series of evolution proposals to bring the language into alignment with its proposed design.

That is, assuming that you're comfortable actually writing a document that satisfies all that. :)

Reply:
I was thinking what the best way would be to express the basic principles i.e. right now I have sort-of made it a part of the proposed changes. Anyway, I have now extracted it out into a separate section, but I would appreciate some advice here as I don't have any language design experience!